types/rdf-js/index.d.ts

```diff
@@ -469,32 +469,32 @@ export interface DatasetCore<Q extends BaseQuad = Quad> {
      */
     match(subject?: Term | null, predicate?: Term | null, object?: Term | null, graph?: Term | null): this;
```

Comment:
I suggest to refer to the terms of `InQuad` here as well.

Reply:
Like this?

```ts
match(
    subject?: PropType<InQuad, 'subject'> | null,
    predicate?: PropType<InQuad, 'predicate'> | null,
    object?: PropType<InQuad, 'object'> | null,
    graph?: PropType<InQuad, 'graph'> | null): this;
```

Reply:
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/40821 was rejected (but still makes sense to me).

Reply:
@tpluscode Something like `InQuad['subject']` is a bit shorter, and consistent with #40821.

@thewilkybarkid Good point, it's the same issue. Let's try to figure this out...
In any case, whatever option we choose, I think it's important to be consistent between `Dataset` and `Store`.

So the question is if incompatible quad subtypes `MyQuad1` and `MyQuad2` are created, should this be allowed?:
```
const source = new MySource<MyQuad1>();
const matchQuad = new SomeQuad<MyQuad2>();
source.match(matchQuad.subject, matchQuad.predicate, matchQuad.object);
```

In hindsight, perhaps it doesn't make much sense to allow this. What do you think @tpluscode @thewilkybarkid?

Reply:
Yea, I think I stand by your comment from #40821. Not because you might have different kinds of quads (not really sold on that still) but because in code you might have a `Term` and declarations will prevent it from being used to match:

```ts
// any term we got from other lib or whatever
let subject: Term
let dataset: Dataset

// restricting match will cause this to fail
dataset.match(term)
```

Reply:
Are there any genuine cases for not knowing what type of `Term` you have?

Reply:
It may just be inaccurate types in other, potential 3rd party code. Limiting the match method would likely force a bunch of `if (node.termType === 'NamedNode')` kind of statement to ensure matching terms. I expect this to quickly go out of hand with little benefit

Reply:
> It may just be inaccurate types in other, potential 3rd party code.

IMO, we should assume that typings will be accurate and correct.
Otherwise we risk moving more in the direction of weakly-typed/typeless JS code.

> would likely force a bunch of if (node.termType === 'NamedNode')

Except for the inaccurate typings case you mentioned, I don't expect this to happen (unless desired by the developer).

Reply:
I'm not sure it's a good idea, and also seems unrelated to the current changes. Agree to try in a separate PR?

Reply:
Sure, a separate PR sounds good.